File: budgetbeacon-back/BudgetBeacon.Core/Models/TransactionTreatment.cs
namespace BudgetBeacon.Core.Models;

public static class TransactionTreatment
{
    public const string Income = "Income";
    public const string Expense = "Expense";
    public const string InternalTransfer = "InternalTransfer";
    public const string SavingsInvestment = "SavingsInvestment";
    public const string Refund = "Refund";
    public const string Adjustment = "Adjustment";

    public static readonly IReadOnlyList<string> Allowed =
    [
        Income,
        Expense,
        InternalTransfer,
        SavingsInvestment,
        Refund,
        Adjustment
    ];

    public static string? Normalize(string? treatment)
    {
        var normalized = treatment?.Trim();

        if (string.IsNullOrWhiteSpace(normalized))
        {
            return null;
        }

        return Allowed.FirstOrDefault(allowed =>
            string.Equals(allowed, normalized, StringComparison.OrdinalIgnoreCase));
    }

    public static string GetDefault(decimal amount, string? category)
    {
        if (string.Equals(category, "Income", StringComparison.OrdinalIgnoreCase) || amount > 0)
        {
            return Income;
        }

        if (string.Equals(category, "Transfers & Adjustments", StringComparison.OrdinalIgnoreCase))
        {
            return InternalTransfer;
        }

        if (string.Equals(category, "Savings & Investments", StringComparison.OrdinalIgnoreCase))
        {
            return SavingsInvestment;
        }

        return Expense;
    }
}

File: budgetbeacon-front/src/features/transactions/transactionTreatment.ts
import type { TransactionTreatment } from "../../types/api";

export const transactionTreatmentOptions: Array<{
  value: TransactionTreatment;
  label: string;
  description: string;
}> = [
  {
    value: "Income",
    label: "Income",
    description: "Counts toward income.",
  },
  {
    value: "Expense",
    label: "Expense",
    description: "Counts toward spending charts.",
  },
  {
    value: "InternalTransfer",
    label: "Internal transfer",
    description: "Excluded from income, expenses and balance.",
  },
  {
    value: "SavingsInvestment",
    label: "Savings / investment",
    description: "Tracked separately from expenses.",
  },
  {
    value: "Refund",
    label: "Refund",
    description: "Reduces spending instead of income.",
  },
  {
    value: "Adjustment",
    label: "Adjustment",
    description: "Excluded from normal analytics.",
  },
];

export const getDefaultTransactionTreatment = (
  amount: number,
  category: string,
): TransactionTreatment => {
  if (category === "Income" || amount > 0) {
    return "Income";
  }

  if (category === "Transfers & Adjustments") {
    return "InternalTransfer";
  }

  if (category === "Savings & Investments") {
    return "SavingsInvestment";
  }

  return "Expense";
};

export const getTransactionTreatmentLabel = (
  treatment: TransactionTreatment,
) =>
  transactionTreatmentOptions.find((option) => option.value === treatment)
    ?.label ?? treatment;
